### PR TITLE
examples/evaluation: add promptiter regression loop

### PR DESCRIPTION
**Pull request by mingri31164**
## What changed

- Added a credential-free deterministic Evaluation Service pipeline that evaluates baseline and candidate prompts on separate train and validation sets.
- Added evidence-based failure attribution, PromptIter-style candidate patches, per-case deltas, configurable acceptance gates, and JSON/Markdown audit reports.
- Added six runnable cases and an explicit overfitting round where training improves while held-out validation regresses and is rejected.
- Added CLI validation so unexpected positional arguments cannot silently trigger an unintended prompt-source write-back.

## Why

Prompt optimization needs held-out regression checks and auditable acceptance criteria before a candidate can be written back. The deterministic runner exercises final responses, tool trajectories, execution traces, cost, and latency without requiring an API key.

## Testing

- `go test -count=1 -race -cover ./promptiter/regression` from `examples/evaluation`: **PASS**, 80.7% statement coverage
- `go vet ./promptiter/regression`: **PASS**
- `go test -count=1 ./...` from `examples/evaluation`: **PASS**, 36 packages
- `.github/scripts/check-examples.sh`: **PASS**, all 28 example modules tidy and build successfully
- Deterministic E2E: validation **0.5000 -> 1.0000**; all five acceptance checks pass
- Overfit regression: train **0.6667 -> 0.8333**, validation **1.0000 -> 0.6667**; candidate rejected with four gate reasons
- Relative and absolute config-path runs produce identical reports except `startedAt`

### 1. Targeted tests, coverage, and repository checks

Targeted tests, coverage, and repository checks

### 2. Held-out validation improvement and acceptance gate

Held-out validation improvement and acceptance gate

### 3. Overfitting rejection and audit artifacts

Overfitting rejection and audit artifacts

## Notes for reviewers

- The public sample contains 3 training cases and 3 held-out validation cases.
- Every failed baseline case has at least one evidence-backed attribution.
- Rejected candidates are never recommended or written back.
- Reports include prompts, metrics, traces, tool trajectories, gate checks, cost, latency, seed, and fake model configuration.

Closes #2003

### CODE REVIEW

**Comment by mingri31164** [comment]
recheck

File: examples/evaluation/promptiter/regression/optimize.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+func candidateTargetsCurrentFailures(
+	config candidateConfig,
+	currentTrain evaluationSummary,
+) bool {
+	if len(config.TargetFailures) == 0 {
+		return true
+	}
+	current := summarizeFailures(currentTrain)
```
Comment (Flash-LHR) [inline examples/evaluation/promptiter/regression/optimize.go]: Using `summarizeFailures` here ignores secondary attributions, so candidates targeting `tool_call_error` are skipped. Match `TargetFailures` against every case attribution.
<details><summary>中文</summary>
这里使用 `summarizeFailures` 会忽略次要归因，因此会跳过以 `tool_call_error` 为目标的候选。请用所有 case 归因匹配 `TargetFailures`。
</details>

File: examples/evaluation/promptiter/regression/evaluate.go [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+	expectedInvocations := expected.Conversation
+	actualResponse := lastResponse(actualInvocations)
+	expectedResponse := lastResponse(expectedInvocations)
+	actualTools := flattenTools(actualInvocations)
+	expectedTools := flattenTools(expectedInvocations)
+	metricResults := result.MetricResults
+	if len(result.EvalCaseResults) == 1 && result.EvalCaseResults[0] != nil {
+		metricResults = result.EvalCaseResults[0].OverallEvalMetricResults
+	}
+	metrics := adaptMetrics(metricResults)
```
Comment (Rememorio) [inline examples/evaluation/promptiter/regression/evaluate.go:337]: **P1: Reject evaluation cases that failed before scoring**

Evaluation Service reports per-case inference and evaluator failures through `Inference.Status`/`ErrorMessage` and the run result's `ErrorMessage` without necessarily failing `Evaluate`. This adapter ignores those fields; the failed case produces no adapted metrics and is therefore omitted from the aggregate score denominator.

With another healthy validation case, the report can show an inflated score even though one case never ran successfully. If that case was already failing in the baseline, or new hard failures are allowed, the gate can accept and write back the candidate. Return an error for inference/evaluation errors (or account for every configured metric as a zero), and add a mixed successful/failed validation-case test.

<details><summary>中文</summary>

**拒绝在评分前已经失败的评测用例**

Evaluation Service 不一定会让整个 `Evaluate` 返回错误，而是可能通过 `Inference.Status`、`ErrorMessage` 或单次运行结果中的 `ErrorMessage` 表示某个用例执行失败。这里没有检查这些字段；失败用例不会产生有效 metric，随后也不会进入总分分母。

只要验证集中还有正常用例，报告就可能给出偏高的分数，即使某个用例根本没有成功完成。如果该用例在 baseline 中本来就失败，或配置允许新增硬失败，候选仍可能通过门禁并被写回。建议遇到推理或评测执行错误时直接终止，或者为每个配置的 metric 明确计零，并增加“正常用例与执行失败用例混合”的验证测试。

</details>

File: examples/evaluation/promptiter/regression/pipeline.go [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+			Total:              totalCosts,
+			TotalLatencyMillis: evaluationLatency(baseline.Train, baseline.Validation) + roundsLatency(rounds),
+		},
+		Rounds: rounds,
+	}
+	if err := writeReports(options.OutputDir, report); err != nil {
+		return optimizationReport{}, err
+	}
+	if options.WritePrompt && report.GateDecision.Accepted {
+		if err := os.WriteFile(config.Inputs.PromptSource, []byte(strings.TrimSpace(selected.Prompt)+"\n"), 0o644); err != nil {
```
Comment (Rememorio) [inline examples/evaluation/promptiter/regression/pipeline.go]: **P1: Replace the prompt source atomically**

`os.WriteFile` truncates the configured prompt source before the replacement is fully written. An `ENOSPC`, I/O failure, or interruption can therefore leave the original user-authored prompt empty or partial even though the pipeline returns an error.

Write to a temporary file in the same directory, preserve the source mode, sync and close it, then rename it over the source. A failure-path test should verify that the original prompt remains intact when the replacement cannot be completed.

<details><summary>中文</summary>

**以原子方式替换提示词源文件**

`os.WriteFile` 会先截断配置中的提示词源文件，再写入新内容。若发生磁盘空间不足、I/O 错误或写入中断，即使流水线最终返回错误，原始提示词也可能已经变成空文件或残缺内容。

建议先在同一目录写入临时文件，保留源文件权限，完成同步和关闭后再通过 rename 原子替换。还应增加失败路径测试，确认替换未完成时原提示词保持不变。

</details>

File: examples/evaluation/promptiter/regression/DESIGN.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+and pass/fail states. The loop does not infer success from optimizer output.
+This makes fake runs and model-backed runs comparable and prevents a candidate
+from bypassing evaluation.
+
+Failure attribution is deterministic and evidence based. It first inspects
+expected and actual knowledge-search calls, structured JSON validity, tool
+names and arguments, trace routing errors, and metric reasons. A generic final
+response mismatch is used only after more specific signals have been checked.
+Every failed case receives at least one category and human-readable evidence;
+the first category is the primary attribution used in summary counts. The full
```
Comment (Rememorio) [inline examples/evaluation/promptiter/regression/DESIGN.md]: **P2: Document that summaries count every attribution**

The current `summarizeFailures` implementation and its test count every attribution attached to a failed case, not only the first category. The generated sample already demonstrates this: category totals can substantially exceed the number of failed cases.

Either restore primary-only summary counting or update this paragraph to explain that summaries count all matched categories and may double-count a case. Otherwise reviewers can misinterpret the audit table as a mutually exclusive root-cause breakdown.

<details><summary>中文</summary>

**说明汇总会统计全部归因类别**

当前 `summarizeFailures` 的实现和测试会统计失败用例上的每一个归因类别，而不是只统计第一个主归因。生成的示例报告已经体现出这一点：各类别数量之和可能明显大于失败用例总数。

建议恢复“只汇总主归因”的逻辑，或者修改本段，明确说明汇总会统计全部命中类别，同一用例可能被重复计数。否则审阅者容易把归因表误解为互斥的根因分布。

</details>

**Review by Rememorio — COMMENTED** [review]
I reviewed the changed lines and left 5 inline comments below. The comments focus on issues worth addressing before merge.

Fresh review found two P1 risks in acceptance/write-back safety and three P2 configuration/documentation issues. The package adds no production exported API. Add focused tests for failed inference cases, zero budgets, negative gain thresholds, and write-back failures; align the attribution design text with the generated summaries.

<details><summary>中文</summary>

我看过这次变更的相关代码，在下面留下 5 条行内评论。评论聚焦在合并前值得处理的问题。

本次审查发现两个涉及验收与写回安全的 P1 问题，以及三个配置和文档一致性的 P2 问题。该示例包没有新增生产导出 API。建议补充推理失败、零预算、负增益阈值和写回失败场景的定向测试，并让归因设计文档与实际汇总逻辑保持一致。

</details>

File: examples/evaluation/promptiter/regression/config.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+		return errors.New("prompt source path is empty")
+	case strings.TrimSpace(config.FakeModel.Name) == "":
+		return errors.New("fake model name is empty")
+	case config.FakeModel.LatencyMillis < 0:
+		return errors.New("fake model latency must not be negative")
+	case config.FakeModel.PromptCostPerMillionTokens < 0:
+		return errors.New("fake model prompt token cost must not be negative")
+	case config.FakeModel.OutputCostPerMillionTokens < 0:
+		return errors.New("fake model output token cost must not be negative")
+	case config.Gate.MaxCriticalScoreDrop < 0:
```
Comment (Rememorio) [inline examples/evaluation/promptiter/regression/config.go:110]: **P2: Reject negative minimum validation gain**

The other gate thresholds are checked for negative values, but `MinValidationScoreGain` is not. A configuration such as `-0.5` explicitly allows a candidate to lose up to half of the validation score while still passing this check; regressions in already-failing or still-passing cases may then reach write-back.

Validate this field as nonnegative and add a configuration test for a negative value.

<details><summary>中文</summary>

**拒绝负数的最小验证增益**

其他门禁阈值都会拒绝负数，但 `MinValidationScoreGain` 没有对应校验。例如配置为 `-0.5` 时，候选即使损失一半验证分数也可能通过该检查；若退化发生在原本已失败或仍未跌破通过阈值的用例上，最终仍可能进入写回流程。

建议要求该字段不小于零，并补充负值配置测试。

</details>

File: examples/evaluation/promptiter/regression/gate.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+			"candidate introduced %d new hard fail(s); allowNewHardFails=%t",
+			delta.NewlyFailed,
+			config.AllowNewHardFails,
+		),
+	)
+
+	criticalPassed, criticalDetail := checkCriticalCases(config, baseline, candidate)
+	addCheck("critical_cases", criticalPassed, criticalDetail)
+
+	costPassed := config.MaxEstimatedCostUSD <= 0 ||
```
Comment (Rememorio) [inline examples/evaluation/promptiter/regression/gate.go]: **P2: Do not treat an explicit zero budget as unlimited**

Both resource checks treat values `<= 0` as disabled, while configuration validation rejects negative values. Consequently there is no way to express a strict zero-cost or zero-tool-call policy: setting either documented maximum to `0` reports a zero limit but always passes the check.

Distinguish an omitted limit from an explicit zero, for example with pointer fields or separate enable flags, and test that a candidate with nonzero usage is rejected under a configured zero budget.

<details><summary>中文</summary>

**不要把显式零预算当作无限制**

两个资源门禁都把 `<= 0` 解释为关闭限制，而配置校验又禁止负数。因此当前无法表达“成本必须为零”或“不得调用工具”：把文档中的最大值显式设为 `0` 时，报告会显示限制为零，但检查始终通过。

建议区分“未配置”与“显式配置为零”，例如使用指针字段或独立启用开关，并增加测试，确保零预算下任何非零资源消耗都会被拒绝。

</details>

File: examples/evaluation/promptiter/regression/optimize.go (hunk 1 above)
Comment (mingri31164) [reply examples/evaluation/promptiter/regression/optimize.go]: Fixed in ca85f1cd. Candidate targeting now scans every attribution on every training case, and TestCandidateTargetsCurrentFailuresMatchesSecondaryAttribution covers a secondary tool_call_error.

File: examples/evaluation/promptiter/regression/pipeline.go [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+	selected.CandidateID = "baseline"
+	rounds := make([]roundAudit, 0, config.MaxRounds)
+	var candidateCosts costSummary
+	evaluatedRounds := 0
+	for _, candidateConfig := range config.Candidates {
+		if evaluatedRounds >= config.MaxRounds {
+			break
+		}
+		if !candidateTargetsCurrentFailures(candidateConfig, current.Train) {
+			continue
```
[the remark this answers is not in the document]
Comment (mingri31164) [reply examples/evaluation/promptiter/regression/pipeline.go:78]: No code change is needed. validateConfig rejects maxRounds <= 0 and an empty candidate list before runPipeline enters the loop, so evaluatedRounds == 0 only means all configured candidates were skipped by failure targeting. The existing validation path keeps this message accurate.

File: examples/evaluation/promptiter/regression/runner.go [hunk 8]
```diff
@@ -0,0 +1,298 @@
+//
+// Tencent is pleased to support the open source community by making trpc-agent-go available.
+//
+// Copyright (C) 2025 Tencent.  All rights reserved.
+//
+// trpc-agent-go is licensed under the Apache License Version 2.0.
+//
```
[the remark this answers is not in the document]
Comment (mingri31164) [reply examples/evaluation/promptiter/regression/runner.go:7]: Fixed in d51af97b. All newly added Go files now use the complete repository header, including the required trailing // line.

File: examples/evaluation/promptiter/regression/evaluate.go (hunk 2 above)
Comment (mingri31164) [reply examples/evaluation/promptiter/regression/evaluate.go:337]: Fixed in bb9b919a. adaptCase now rejects inference and evaluator errors, nil or not-evaluated metrics, and missing run results; adaptEvaluation also rejects missing and duplicate cases. Mixed success/failure and missing-case tests ensure failed executions cannot disappear from the denominator.

File: examples/evaluation/promptiter/regression/pipeline.go (hunk 3 above)
Comment (mingri31164) [reply examples/evaluation/promptiter/regression/pipeline.go]: Fixed in bb9b919a. Accepted prompts are written to a same-directory temporary file with the source mode preserved, then synced, closed, and atomically renamed. Tests cover successful mode preservation and an injected rename failure that leaves the original prompt intact and removes the temporary file.

File: examples/evaluation/promptiter/regression/DESIGN.md (hunk 4 above)
Comment (mingri31164) [reply examples/evaluation/promptiter/regression/DESIGN.md]: Fixed in bb9b919a. DESIGN now states that summaries count every matched attribution, so one case may contribute to multiple categories and totals may exceed the failed-case count.

File: examples/evaluation/promptiter/regression/config.go (hunk 5 above)
Comment (mingri31164) [reply examples/evaluation/promptiter/regression/config.go:110]: Fixed in bb9b919a. Configuration validation now rejects a negative minValidationScoreGain, with a focused configuration test covering the invalid value.

File: examples/evaluation/promptiter/regression/gate.go (hunk 6 above)
Comment (mingri31164) [reply examples/evaluation/promptiter/regression/gate.go]: Fixed in bb9b919a. Budget fields are now pointers: omission disables a limit, while an explicit 0 enforces strict zero usage. Tests cover explicit-zero rejection, including a scoreEpsilon-sized cost, and omitted-budget behavior.

File: examples/evaluation/promptiter/regression/evaluate.go (hunk 2 above)
Comment (Rememorio) [reply examples/evaluation/promptiter/regression/evaluate.go:337]: Fixed. The adapter now rejects failed inference/evaluation runs and missing or duplicate cases, and the mixed-result tests cover the denominator regression.

<details><summary>中文</summary>

已修复。适配器现在会拒绝推理或评测执行失败以及缺失、重复用例，混合结果测试也覆盖了分母被缩小的问题。

</details>

File: examples/evaluation/promptiter/regression/pipeline.go (hunk 3 above)
Comment (Rememorio) [reply examples/evaluation/promptiter/regression/pipeline.go]: Fixed. Replacement now uses a synced same-directory temporary file, and the injected rename-failure test confirms that the original prompt remains intact.

<details><summary>中文</summary>

已修复。替换操作现在使用完成同步的同目录临时文件，注入重命名失败的测试也确认原提示词保持完整。

</details>

File: examples/evaluation/promptiter/regression/DESIGN.md (hunk 4 above)
Comment (Rememorio) [reply examples/evaluation/promptiter/regression/DESIGN.md]: Fixed. The design now explicitly states that all matched attributions are counted and totals may exceed the failed-case count.

<details><summary>中文</summary>

已修复。设计文档现已明确说明所有命中的归因都会计数，因此汇总数可能超过失败用例数。

</details>

File: examples/evaluation/promptiter/regression/config.go (hunk 5 above)
Comment (Rememorio) [reply examples/evaluation/promptiter/regression/config.go:110]: Fixed. Negative `minValidationScoreGain` values are now rejected and covered by a focused configuration test.

<details><summary>中文</summary>

已修复。负数的 `minValidationScoreGain` 现在会被拒绝，并已有针对性的配置测试。

</details>

File: examples/evaluation/promptiter/regression/gate.go (hunk 6 above)
Comment (Rememorio) [reply examples/evaluation/promptiter/regression/gate.go]: Fixed. Pointer-backed limits now distinguish omission from explicit zero, and tests cover both strict-zero and omitted-budget behavior.

<details><summary>中文</summary>

已修复。指针形式的限制现在可以区分未配置和显式零值，测试也覆盖了严格零预算及省略预算的行为。

</details>

File: examples/evaluation/promptiter/regression/pipeline.go [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+		},
+		CostLatency: costLatencySummary{
+			Baseline:           baselineCosts,
+			Candidates:         candidateCosts,
+			Total:              totalCosts,
+			TotalLatencyMillis: evaluationLatency(baseline.Train, baseline.Validation) + roundsLatency(rounds),
+		},
+		Rounds: rounds,
+	}
+	if err := writeReports(options.OutputDir, report); err != nil {
```
Comment (Rememorio) [inline examples/evaluation/promptiter/regression/pipeline.go:211]: **P1: Reject report targets that alias pipeline inputs**

`writeReports` writes two fixed filenames without checking whether either target is also the config, prompt source, metrics file, or an eval set. For example, choosing an output directory that contains a prompt named `optimization_report.md` overwrites that prompt even with `WritePrompt=false`, contradicting the audit-only guarantee and potentially destroying user-authored input.

Resolve both report targets before writing and reject aliases against every input path; account for existing symlink or hard-link aliases with `os.SameFile`. A focused audit-only test should verify that a collision returns an error and leaves the input intact.

<details><summary>中文</summary>

**拒绝与流水线输入重合的报告路径**

`writeReports` 会直接写入两个固定文件名，却没有检查目标文件是否同时也是配置、提示词源、metrics 或 eval set。比如输出目录中恰好存在名为 `optimization_report.md` 的提示词文件时，即使 `WritePrompt=false`，该文件仍会被报告覆盖。这既违背“只审计、不修改源文件”的承诺，也可能破坏用户输入。

建议在任何写入前解析两个报告目标，并与所有输入路径进行冲突检查；对已存在的符号链接或硬链接可使用 `os.SameFile` 判断。还应增加一个只审计场景的测试，确认路径冲突会返回错误且原输入保持不变。

</details>

File: examples/evaluation/promptiter/regression/README.md [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+Supported intents are `lookup`, `structured`, `knowledge`, and `direct`.
+Hidden or custom cases can use the same schema without changing Go code.
+
+## Production Adapter
+
+The deterministic candidate generator preserves PromptIter's patch/profile
+contract but replaces its LLM-backed backward, aggregation, and optimizer
+stages. A production integration can obtain candidates from
+`evaluation/workflow/promptiter/engine`, then pass each candidate prompt and
```
Comment (Rememorio) [inline examples/evaluation/promptiter/regression/README.md:110]: **P2: Clarify the non-importable production reuse path**

A production integration cannot pass results into “the same delta, gate, and report functions” as written: this example is `package main`, and those functions and their parameter types are all unexported. There is no importable package elsewhere in the repository providing this contract.

If this section describes a pattern to copy or adapt, say that explicitly. If direct reuse is intended, move the coherent reusable surface into an importable package and document and test its exported contract before promising it here.

<details><summary>中文</summary>

**澄清当前无法直接导入的生产复用路径**

生产集成目前无法按文档所述把结果传给“同一组 delta、gate 和 report 函数”：该示例属于 `package main`，相关函数及其参数类型也全部未导出，仓库中没有其他可导入的包提供这套契约。

如果这里只是希望用户参考或改写这一模式，应在文档中明确说明；如果确实要支持直接复用，则需要先把完整且稳定的能力放入可导入包，为导出契约补齐文档和测试，再在此处作出复用承诺。

</details>

File: examples/evaluation/promptiter/regression/evaluate.go [hunk 11]
```diff
@@ -0,0 +1,489 @@
...
+	resultManager := evalresultinmemory.New()
+	for _, evalSet := range []*evalset.EvalSet{train, validation} {
+		if err := addEvalSet(ctx, evalSetManager, config.AppName, evalSet); err != nil {
+			return nil, err
+		}
+		for _, evalMetric := range metrics {
+			if err := metricManager.Add(ctx, config.AppName, evalSet.EvalSetID, evalMetric); err != nil {
+				return nil, fmt.Errorf(
+					"add metric %q to eval set %q: %w",
+					evalMetric.MetricName,
```
Comment (Rememorio) [inline examples/evaluation/promptiter/regression/evaluate.go:131]: **P2: Reject null metric entries before registration**

A metrics file containing `[null]` passes `validateEvaluationInputs` because the slice is non-empty. `metricManager.Add` then returns `metric is nil`, but this error path dereferences `evalMetric.MetricName`, so malformed external input crashes the CLI instead of returning a validation error.

Validate every metric entry as non-nil before constructing the runtime, and add a loading test for a null list element. The formatter should also avoid dereferencing the value when registration fails.

<details><summary>中文</summary>

**在注册前拒绝空 metric 条目**

metrics 文件若包含 `[null]`，会因为切片长度不为零而通过 `validateEvaluationInputs`。随后 `metricManager.Add` 返回 `metric is nil`，但这里在组装错误信息时又访问 `evalMetric.MetricName`，导致 CLI 因外部输入格式错误而直接 panic，而不是正常返回校验错误。

建议在创建 runtime 前逐项检查 metric 不为 nil，并补充包含空元素的加载测试；注册失败的错误处理也不应再次解引用该值。

</details>

**Review by Rememorio — COMMENTED** [review]
I reviewed the changed lines and left 3 inline comments below. The comments focus on issues worth addressing before merge.

Fresh review found three actionable issues: report targets can overwrite pipeline inputs, a null metric entry can panic during error handling, and the documented production reuse path is not importable. All five previous review concerns are fixed in the current head. No exported production symbols were added; the only API-boundary issue is the reuse claim in the documentation.

<details><summary>中文</summary>

我看过这次变更的相关代码，在下面留下 3 条行内评论。评论聚焦在合并前值得处理的问题。

本轮复审发现三个需要处理的问题：报告文件可能覆盖流水线输入、空 metric 条目会在错误处理时触发 panic，以及文档描述的生产复用路径实际无法导入。此前五条评审意见在当前提交中均已修复。本次没有新增可导出的生产 API；唯一的 API 边界问题是文档中的复用承诺与实现不符。

</details>

**Comment by mingri31164** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by liuzengh** [comment]
感谢你对本 Issue 的贡献！

犀牛鸟活动的 Issue 提交阶段已经结束，本 PR 将关闭。再次感谢你的参与与付出！
